apps/emqx_rule_engine/src/emqx_rule_actions.erl

```diff
@@ -232,3 +243,70 @@ format_pub_props(UserPropertiesTks, Selected, Env) ->
                 replace_simple_var(UserPropertiesTks, Selected, #{})
         end,
     #{'User-Property' => UserProperties}.
+
+format_mqtt_properties(MQTTPropertiesTemplate, Selected, Env) ->
+    #{metadata := #{rule_id := RuleId}} = Env,
+    MQTTProperties0 =
+        maps:fold(
+            fun(K, Template, Acc) ->
+                try
+                    V = emqx_placeholder:proc_tmpl(Template, Selected),
+                    Acc#{K => V}
+                catch
+                    Kind:Error ->
+                        ?SLOG(
+                            error,
+                            #{
+                                msg => "bad_mqtt_prop_value",
+                                rule_id => RuleId,
+                                reason => {Kind, Error},
+                                property => K,
+                                selected => Selected
+                            }
+                        ),
+                        Acc
+                end
+            end,
+            #{},
+            MQTTPropertiesTemplate
+        ),
+    coerce_properties_values(MQTTProperties0, Env).
+
+ensure_int(B) when is_binary(B) -> binary_to_integer(B);
+ensure_int(I) when is_integer(I) -> I.
+
+coerce_properties_values(MQTTProperties, #{metadata := #{rule_id := RuleId}}) ->
+    PublishProperties =
+        maps:from_list([
+            {K, T}
+         || {_Id, {K, T, Tags}} <- maps:to_list(emqtt_props:all()),
+            is_list(Tags) andalso lists:member(?PUBLISH, Tags)
+        ]),
```

Comment:
since we are enumerating all possible types in schema anyway.
for best readability and performance, we can enumerate the encoding types for them as function clauses.

Reply:
```
encode_properties_value('Payload-Format-Indicator', V) -> ensure_int(V);
```